R/helper.R

```diff
@@ -19,3 +19,39 @@ get_by_id = function(xs, id) {
   }
   return(NULL)
 }
+
+
+#' @description calculate pareto front of an all-numeric data frame (or matrix)
+#'   also works for >2D pareto front
+#' @example data_frame = data.frame(c(1, 0.25, 0.75, 0), c(0, 0.25, 0.75, 1))
+#'   calculate_pareto_front(data_frame)
+#'   data_frame = data.frame(c(1,1,0,1), c(1,1,0,0), c(1,0,1,0), c(1,1,1,1))
+#'   calculate_pareto_front(data_frame, maximize = FALSE)
+calculate_pareto_front = function(data_frame, maximize = TRUE) {
+
+  assert_data_frame(data_frame, types = "numeric")
+  assert_logical(maximize)
+
+  # sort the whole data.frame and go stepwise through the columns whenever
+  # entries are tied
+  sorted = do.call(order, c(as.list(data_frame), decreasing = maximize))
+  sorted_data = data_frame[sorted, ]
+
+# function for cummulative min or max of vector
+  cummaxmin = function(x, m) if (m) cummax(x) else cummin(x)
+   
+  # flag non-violations of strict monotony in each column
+  bool_matrix = mapply(function(x, m) !duplicated(cummaxmin(x, m)), sorted_data, maximize)
+  # check for any flag in each row
+  front = apply(bool_matrix, 1, any)
+  # get original indices
+  revert_sort = as.numeric(rownames(sorted_data))
```

Comment:
To restore the order, use `order(sorted)`.

Reply:
maybe i did something wrong, but it didnt work out when i tried it your way

Reply:
```r
x1 = runif(10)
ord = order(x1)
x2 = x1[ord] # x2 is sorted x1
x3 = x2[order(ord)] # restore order of x1
all(x1 == x3)
```